rust/parquet/src/reader/page_iterator.rs

```diff
@@ -0,0 +1,124 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Provides interface for page iterator and implementation.
+
+use std::rc::Rc;
+
+use crate::column::page::PageReader;
+use crate::file::reader::FileReader;
+use crate::schema::types::ColumnDescPtr;
+use crate::schema::types::SchemaDescPtr;
+
+use crate::errors::check_in_range;
+use crate::errors::Result;
+
+/// PageIterator is used to decouple parquet row group scanning from record readers.
+pub trait PageIterator: Iterator<Item = Result<Box<PageReader>>> {
+    /// Get schema of parquet file.
+    fn schema(&mut self) -> Result<SchemaDescPtr>;
+
+    /// Get column schema of this page iterator.
+    fn column_schema(&mut self) -> Result<ColumnDescPtr>;
+}
+
+/// Implementation for parquet file.
+pub struct FilePageIterator {
+    column_index: usize,
+    row_group_indices: Box<Iterator<Item = usize>>,
+    file_reader: Rc<FileReader>,
+}
+
+impl FilePageIterator {
+    /// Create page iterator from parquet file reader.
+    pub fn new(
+        column_index: usize,
+        row_group_indices: Box<Iterator<Item = usize>>,
```

Comment:
nit: can this parameter be optional and in default the iterator will span all row groups?

Reply:
Good point. Fixed in new PR.